ethcore/engines/authority-round/src/util.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utility functions.
+//!
+//! Contains small functions used by the AuRa engine that are not strictly limited to that scope.
+
+use std::fmt;
+
+use client_traits::EngineClient;
+use common_types::ids::BlockId;
+use ethabi;
+use ethereum_types::Address;
+
+/// A contract bound to a client and block number.
+///
+/// A bound contract is a combination of a `Client` reference, a `BlockId` and a contract `Address`.
+/// These three parts are enough to call a contract's function; return values are automatically
+/// decoded.
+pub struct BoundContract<'a> {
+	client: &'a dyn EngineClient,
+	block_id: BlockId,
+	contract_addr: Address,
+}
+
+/// Contract call failed error.
+#[derive(Debug)]
+pub enum CallError {
+	/// The call itself failed.
+	CallFailed(String),
+	/// Decoding the return value failed or the decoded value was a failure.
+	DecodeFailed(ethabi::Error),
+	/// The passed in client reference could not be upgraded to a `BlockchainClient`.
+	NotFullClient,
+}
+
+impl<'a> fmt::Debug for BoundContract<'a> {
+	fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+		fmt.debug_struct("BoundContract")
+			.field("client", &(self.client as *const dyn EngineClient))
+			.field("block_id", &self.block_id)
+			.field("contract_addr", &self.contract_addr)
+			.finish()
+	}
+}
+
+impl<'a> BoundContract<'a> {
+	/// Create a new `BoundContract`.
+	#[inline]
+	pub fn bind(client: &dyn EngineClient, block_id: BlockId, contract_addr: Address) -> BoundContract {
+		BoundContract {
+			client,
+			block_id,
+			contract_addr,
+		}
+	}
+
+	/// Perform a function call to an ethereum machine that doesn't create a transaction or change the state.
+	///
+	/// Runs a constant function call on `client`. The `call` value can be serialized by calling any
```

Comment:
I think I've asked this somewhere else already but I do not remember the answer, apologies. What is meant by a "constant function" in this context? Is it "constant" because it does not create txs and thus does not alter state? If yes, I suggest being explicit: "Runs a non-state altering function call…" (or something to that effect).

Reply:
Yes, exactly. I'll expand the comment.

Reply:
The comment was expanded in https://github.com/poanetwork/parity-ethereum/commit/4799aeb3d11bac2910ff04e5714631a541d92aec#diff-a4faad7e2fe2f519b612d48dfa0f8607.